### PR TITLE
Don't rebuild dev images

### PR DESCRIPTION
**Pull request by lhoestq**
To improve the local dev environment and let us iterate quicker:
1. I created `dev.Dockerfile` for the admin and api services
2. I improved the dev-docker-compose to use volumes for libcommon, admin and api
3. I removed a step in the `Dockerfile`s
 that was unecessarilly reinstalling python (and possibly at another version than 3.9.15)
4. I updated poetry (1.3.2 can't install `torch` in a linux container on mac m2) and fixed the `tensorflow` dependency to have it working locally on a mac with amd chip (this impacts the dev images but also the prod images, but I think it's ok, it was introduced in https://github.com/huggingface/datasets-server/pull/823)

With those changes I don't need to rebuild images anymore when changing code in libcommon/admin/api/worker etc.

The image is rebuilt only when the dependencies change.

### CODE REVIEW

File: services/worker/dev.Dockerfile [hunk 1]
```diff
@@ -10,13 +10,13 @@ ENV PYTHONFAULTHANDLER=1 \
     PIP_DEFAULT_TIMEOUT=100 \
     POETRY_NO_INTERACTION=1 \
     # Versions:
-    POETRY_VERSION=1.3.2 \
+    POETRY_VERSION=1.4.0 \
```
Comment (severo) [inline services/worker/dev.Dockerfile:13]: would it make sense to also upgrade in the other docker images?

File: services/worker/pyproject.toml [hunk 2]
```diff
@@ -31,7 +31,7 @@ pypdf2 = "^3.0.1"
 python = "3.9.15"
 rarfile = "^4.0"
 scikit-learn = "^1.2.1"
-tensorflow-cpu = {version = "^2.11.0", markers = "sys_platform != 'darwin' or platform_machine != 'arm64'"}
+tensorflow = {version = "^2.11.0", markers = "sys_platform != 'darwin' or platform_machine != 'arm64'"}
```
Comment (severo) [inline services/worker/pyproject.toml:34]: The docker images will include 2GB of useless binaries (for GPU) if we change back to tensorflow. See https://github.com/huggingface/datasets-server/pull/680/files.

I'm not sure what would be the best compromise to be able to work on MacOS locally and to have the smallest docker images possible.

**Review by severo — APPROVED** [review]
Please review the size of the worker image after changing tensorflow-cpu to tensorflow... It should add binaries for GPU which are not used in prod.

File: services/worker/pyproject.toml (hunk 2 above)
Comment (lhoestq) [reply services/worker/pyproject.toml:34]: hmm let me try alternatives then

**Comment by lhoestq** [comment]
I updated poetry in the other places, and added tensorflow-aarch64 for linux containers running on amd macbooks. This way the prod images keep using `tensorflow-cpu`:)

**Comment by severo** [comment]
Thanks!!

**Pull request merged**